docs/src/configuration.md
# Configuration

`locast2tuner` parameters can be specified as either command line arguments, as environment variables, or in a [TOML](https://github.com/toml-lang/toml) configuration file that can be specified using the `--config` argument.

> The precedence of parameters is **command line arguments** > **environment variables** > **config file**
## Command line arguments
In order to specify command line arguments, use `--argument value` (e.g. `--override_zipcodes 90210,30301`). Note that underscores are retained.

## Environment variables
Parameters can also be specified using environment variables. Environment variables are prefixed with `l2t_` and should be exported in the format `l2t_argument="value"` (e.g. `export l2t_override_zipcodes="90210,30301"`). Note that underscores and lowercase are retained.

For boolean flags (like `l2t_multiplex`), use `true` or `false` as values.

## Configuration file
The configuration file format is:

```toml
string_option = "<value1>"
flag = <true/false>
numerical_option = <number>
list_option = ["<value1>", "<value2>"]
```

Example:
```toml
username = "<Locast username>"
password = "<Locast password>"
verbose = 2
multiplex = true
override_zipcodes = ["85355", "90210"]
```

## Configuration options
Please have a look at the [example config file](https://raw.githubusercontent.com/wouterdebie/locast2tuner/main/assets/config.example)

Option | Description | Default
- | - | -
username _(required)_   | Locast.org username
password _(required)_   | Locast.org password
bind_address            | Address of the interface to bind to. To bind to all interfaces, use 0.0.0.0 | 127.0.0.1
cache_dir               | Cache data location | `$HOME/.locast2tuner`
cache_timeout           | How often (in seconds) the station cache is refreshed | 3600
days                    | Days of EPG data to fetch | 8, which is the maximum of data locast.org provides
device_firmware         | Device firmware that is reported to Plex or Emby | homerun3_atsc
device_model            | Device model that is reported to Plex or Emby | HDHR3-US
device_version          | Device version that is reported to Plex or Emby | 20170612
disable_station_cache   | Disable caching of station information. By default `locast2tuner` caches station information for an hour (see `cache_timeout`). By disabling the cache, every request for station information will lead to a call to locast.org. Normally you shouldn't have to disable the cache | false
disable_donation_check  | Disable the donation check. This doesn't mean you can watch without a donation, but the donation check fails for Locast Cares accounts | false
logfile                 | Log to a specific file | By default `locast2tuner` will not log to a file
multiplex               | Normally, when you override multiple zip codes, `locast2tuner` starts multiple instances (see "bind_address"), but with "multiplex = true", stations from multiple locations will be available through a single instance | false
no_tvc_guide_station    | Don't include `tvc_guide_station` in `tuner.m3u`. Having this field sometimes breaks things in Channels DVR. | false
override_cities         | Cities to override the location. Please see [locast.org](https://www.locast.org/dma) for a current map of the supported regions. This should be a pipe separated list with cities and states. E.g. `--override_cities "Los Angeles, CA\|Portland, OR"`| Unset. `locast2tuner` will use your external IP to determine your location
override_zipcodes       | Zip codes to override the location. Please see [locast.org](https://www.locast.org/dma) for a current map of the supported regions. This should be a comma separated list. E.g. `--override_zipcodes "90210,33101"`| Unset. `locast2tuner` will use your external IP to determine your location
port                    | TCP port to bind to. The default is 6077. In case you override muliple zip codes, `locast2tuner` will bind to multiple ports, starting at the number specified below (or default 6077). Any additional instance will bind to a port incremented by 1. E.g. if you override 3 zip codes, 3 instances will be started and bound to 6077, 6078 and 6079. In order to only use one instance, use `multiplex` | 6077
quiet                   | Don't output anything to the terminal | false
random_zipcode      | When `--override_cities` is used, `locast2tuner` looks up a list of valid zip codes for each city and will pick the first valid zip code, with `--random_zipcode` a random valid zip code for the city specified will be picked. | false
remap                   | Remap channel numbers when `multiplexing`. In case you override multiple zip codes, Emby and Plex will sort channels by channel number, which means channels from different locations might be intermingled. In order circumvent this, you can use "remap = true". This causes `locast2tuner` to rewrite the channel number based on the amount of instances there are. Locast will remap a "channel_number" to "channel_number + 100 * instance_number", where the instance_number starts at 0. E.g. you override 3 zip codes, then the channels from the first location will be untouched (since 100*0 == 0), the stations for the second location will start at 100 (e.g. 2.1 CBS becomes 102.1 CBS) and the stations for the third location will start at 200 (e.g. 13.2 WWFF becomes 213.2 WWFF). Note that `multiplex` has to be enabled! | false
remap_file              | File that can be used to do a custom remap. More info can be found [here](advanced/remapping.md). Note that `multiplex` has to be enabled! | Unset
skip_hls                | Instead of using hls.locastnet.org, use the proxy closer to the destination | false

rust_backtrace          | Enable RUST_BACKTRACE=1. In error logs, you might see "run with `RUST_BACKTRACE=1` environment variable to display a backtrace". Instead of adding the environment variable, you can enable this behavior with `rust_backtrace` | false
syslog                  | Log through syslogd | false
tuner_count             | The amount of tuners that is communicated to Plex. This will limit the amount of streams that Plex will Note that this is not a limitation in `locast2tuner` | 16
verbose                 | Verbosity. 0 = Info, 1 = Info + HTTP request lgos, 2 = Debug, 3 = Trace. In error logs, you might see "run with `RUST_BACKTRACE=1` environment variable to display a backtrace". Setting the verbosity to 2 or 3 will also include the backtrace | 0

## Displaying running config
You can display your running config (which could be a combination of a config file and command line parameters) by opening the `/config` path (e.g. `http://127.0.0.1:6077/config`). Normally the password is obfuscated, but if you add the query parameter `show_password` (e.g. `http://127.0.0.1:6077/config?showpass`), the password will become visible.


src/config.rs
use clap_conf::convert::Holder;
use clap_conf::convert::Localizer;
use clap_conf::env::Enver;
use clap_conf::*;
use serde::Serialize;
use simple_error::SimpleError;
use std::fs;
use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::path::PathBuf;
use toml::Value;
use uuid::Uuid;
#[derive(Default, Debug, Serialize, Clone)]
pub struct Config {
    pub logfile: Option<String>,
    pub remap_file: Option<String>,
    pub bind_address: String,
    pub cache_directory: PathBuf,
    pub cache_timeout: u64,
    pub days: u8,
    pub device_firmware: String,
    pub device_model: String,
    pub device_version: String,
    pub disable_station_cache: bool,
    pub disable_donation_check: bool,
    pub multiplex: bool,
    pub no_tvc_guide_station: bool,
    pub override_cities: Option<Vec<String>>,
    pub override_zipcodes: Option<Vec<String>>,
    pub password: String,
    pub port: u16,
    pub quiet: bool,
    pub random_zipcode: bool,
    pub remap: bool,
    pub skip_hls: bool,
    pub rust_backtrace: bool,
    pub syslog: bool,
    pub tuner_count: u8,
    pub username: String,
    #[serde(skip_serializing)]
    pub uuid: String,
    pub verbose: u8,
}
impl Config {
    pub fn from_args_and_file() -> Result<Config, SimpleError> {
        let clap = clap_app!(
            locast2tuner=>
                (version: crate_version!())
                (author: "Wouter de Bie")
                (about: "Locast to tuner")
                (@arg bind_address: -b --bind_address +takes_value "Bind address (default: 127.0.0.1)")
                (@arg cache_dir: --cache_dir +takes_value "Cache directory (default: $HOME/.locast2tuner)")
                (@arg cache_timeout: --cache_timeout +takes_value "Cache timeout (default: 3600)")
                (@arg config: -c --config +takes_value "Config File") //allow clap_conf config loader to work
                (@arg days: -d --days +takes_value "Nr. of days to get EPG data for (default: 8)")
                (@arg device_firmware: --device_firmware +takes_value "Device firmware (default: hdhomerun3_atsc)")
                (@arg device_model: --device_model +takes_value "Device model (default: HDHR3-US)")
                (@arg device_version: --device_version +takes_value "Device version (default: 20170612)")
                (@arg disable_station_cache: --disable_station_cache "Disable stations cache")
                (@arg disable_donation_check: --disable_donation_check "Disable the donation check (use for Locast Cares accounts")
                (@arg multiplex: -m --multiplex "Multiplex devices")
                (@arg override_zipcodes: -z --override_zipcodes +takes_value "Override locations using zipcodes")
                (@arg override_cities: --override_cities +takes_value "Override locations using cities")
                (@arg password: -P --password +takes_value "Locast password")
                (@arg port: -p --port +takes_value "Bind TCP port (default: 6077)")
                (@arg remap: -r --remap "Remap channels when multiplexed. Requires multiplex!")
                (@arg rust_backtrace: --rust_backtrace "Enable RUST_BACKTRACE=1")
                (@arg syslog: --syslog "Log to syslogd")
                (@arg quiet: --quiet "Don't log to terminal")
                (@arg tuner_count: --tuner_count +takes_value "Tuner count (default: 16)")
                (@arg username: -U --username +takes_value "Locast username")
                (@arg verbose: -v --verbose +takes_value "Verbosity (default: 0)")
                (@arg logfile: -l --logfile +takes_value "Log file location")
                (@arg remap_file: --remap_file +takes_value "Remap file location. Requires muliplex!")
                (@arg no_tvc_guide_station: --no_tvc_guide_station "Don't show no_tvc_guide_station in tuner.m3u")
                (@arg skip_hls: --skip_hls "Skip hls.locast.org, but use endpoints that are close to the broadcast")
                (@arg random_zipcode: --random_zipcode "Randomize city zip codes when using override_cities")

        )
        .get_matches();

        let mut conf = Self::default();
        let cfg = clap_conf::with_toml_env(&clap, &["/etc/locast2tuner/config"]);

        conf.username = cfg
            .grab()
            .arg("username")
            .env("l2t_username")
            .conf("username")
            .done()
            .expect("Username required");

        conf.password = cfg
            .grab()
            .arg("password")
            .env("l2t_password")
            .conf("password")
            .done()
            .expect("Password required");

        conf.bind_address = cfg
            .grab()
            .arg("bind_address")
            .env("l2t_bind_address")
            .conf("bind_address")
            .def("127.0.0.1");

        conf.port = cfg
            .grab()
            .arg("port")
            .env("l2t_port")
            .conf("port")
            .t_def::<u16>(6077);

        conf.verbose = cfg
            .grab()
            .arg("verbose")
            .env("l2t_verbose")
            .conf("verbose")
            .t_def::<u8>(0);

        conf.multiplex = cfg.bool_flag("multiplex", Filter::Arg)
            || env_true_flag(&cfg, "l2t_multiplex")
            || cfg.bool_flag("multiplex", Filter::Conf);

        conf.no_tvc_guide_station = cfg.bool_flag("no_tvc_guide_station", Filter::Arg)
            || env_true_flag(&cfg, "l2t_no_tvc_guide_station")
            || cfg.bool_flag("no_tvc_guide_station", Filter::Conf);

        // First check if there's a comma-separated list from the command line
        conf.override_zipcodes = match cfg.grab().arg("override_zipcodes").done() {
            Some(o) => Some(o.split(',').map(|x| x.to_owned()).collect()),
            // Otherwise check for a comma-separated list from env variables
            None => match cfg.grab().env("l2t_override_zipcodes").done() {
                Some(eo) => Some(eo.split(',').map(|x| x.to_owned()).collect()),
                // If nothing, get from config
                None => cfg
                    .grab_multi()
                    .conf("override_zipcodes")
                    .done()
                    .map(|o| o.collect()),
            },
        };

        // First check if there's a comma-separated list from the command line
        conf.override_cities = match cfg.grab().arg("override_cities").done() {
            Some(o) => Some(o.split('|').map(|x| x.to_owned()).collect()),
            // Otherwise check for a comma-separated list from env variables
            None => match cfg.grab().env("l2t_override_cities").done() {
                Some(eo) => Some(eo.split('|').map(|x| x.to_owned()).collect()),
                // If nothing, get from config
                None => cfg
                    .grab_multi()
                    .conf("override_cities")
                    .done()
                    .map(|o| o.collect()),
            },
        };

        conf.random_zipcode = cfg.bool_flag("random_zipcode", Filter::Arg)
            || env_true_flag(&cfg, "l2t_random_zipcode")
            || cfg.bool_flag("random_zipcode", Filter::Conf);

        conf.tuner_count = cfg
            .grab()
            .arg("tuner_count")
            .env("l2t_tuner_count")
            .conf("tuner_count")
            .t_def::<u8>(16);

        conf.device_model = cfg
            .grab()
            .arg("device_model")
            .env("l2t_device_model")
            .conf("device_model")
            .def("HDHR3-US");

        conf.device_firmware = cfg
            .grab()
            .arg("device_firmware")
            .env("l2t_device_firmware")
            .conf("device_firmware")
            .def("hdhomerun3_atsc");

        conf.device_version = cfg
            .grab()
            .arg("device_version")
            .env("l2t_device_version")
            .conf("device_version")
            .def("20170612");

        conf.disable_station_cache = cfg.bool_flag("disable_station_cache", Filter::Arg)
            || env_true_flag(&cfg, "l2t_disable_station_cache")
            || cfg.bool_flag("disable_station_cache", Filter::Conf);

        conf.disable_donation_check = cfg.bool_flag("disable_donation_check", Filter::Arg)
            || env_true_flag(&cfg, "l2t_disable_donation_check")
            || cfg.bool_flag("disable_donation_check", Filter::Conf);

        conf.syslog = cfg.bool_flag("syslog", Filter::Arg)
            || env_true_flag(&cfg, "l2t_syslog")
            || cfg.bool_flag("syslog", Filter::Conf);

        conf.quiet = cfg.bool_flag("quiet", Filter::Arg)
            || env_true_flag(&cfg, "l2t_quiet")
            || cfg.bool_flag("quiet", Filter::Conf);

        conf.cache_timeout = cfg
            .grab()
            .arg("cache_timeout")
            .env("l2t_cache_timeout")
            .conf("cache_timeout")
            .t_def::<u64>(3600);

        conf.days = cfg
            .grab()
            .arg("days")
            .env("l2t_days")
            .conf("days")
            .t_def::<u8>(8);

        conf.remap = cfg.bool_flag("remap", Filter::Arg)
            || env_true_flag(&cfg, "l2t_remap")
            || cfg.bool_flag("remap", Filter::Conf);

        conf.rust_backtrace = cfg.bool_flag("rust_backtrace", Filter::Arg)
            || env_true_flag(&cfg, "l2t_rust_backtrace")
            || cfg.bool_flag("rust_backtrace", Filter::Conf);

        conf.logfile = cfg
            .grab()
            .arg("logfile")
            .env("l2t_logfile")
            .conf("logfile")
            .done();

        conf.remap_file = cfg
            .grab()
            .arg("remap_file")
            .env("lt2_remap_file")
            .conf("remap_file")
            .done();

        conf.skip_hls = cfg.bool_flag("skip_hls", Filter::Arg)
            || env_true_flag(&cfg, "l2t_skip_hls")
            || cfg.bool_flag("skip_hls", Filter::Conf);

        let default_cache_dir = dirs::home_dir().unwrap().join(Path::new(".locast2tuner"));

        let cache_directory_name = cfg
            .grab()
            .arg("cache_dir")
            .env("l2t_cache_dir")
            .conf("cache_dir")
            .def(default_cache_dir.to_str().unwrap());

        let cache_directory = create_cache_directory(cache_directory_name);

        conf.uuid = load_uuid(&cache_directory).unwrap();

        conf.cache_directory = cache_directory;

        if (conf.remap || conf.remap_file.is_some()) && !conf.multiplex {
            panic!("In order to use remap or remap_file, multiplex has to be enabled!");
        }

        Ok(conf)
    }
}

// Create the cache directory
fn create_cache_directory(name: String) -> PathBuf {
    let cache_dir = Path::new(&name).to_path_buf();
    if !cache_dir.exists() {
        fs::create_dir(cache_dir.as_path())
            .expect(&format!("Unable to create directory {:?}", cache_dir)[..]);
    }
    cache_dir
}

// Load the UUID from cache directory if exists
fn load_uuid(cache_directory: &Path) -> Result<String, Box<dyn std::error::Error>> {
    let uid_file = cache_directory.join(Path::new("uuid"));
    let uuid = match uid_file.exists() {
        true => fs::read_to_string(uid_file)?,
        false => generate_and_store_uid(uid_file),
    };

    Ok(uuid)
}

// Generate UUID and store it in the supplied path
fn generate_and_store_uid(path: PathBuf) -> String {
    let new_uuid = Uuid::new_v4().to_string();

    File::create(&path)
        .unwrap()
        .write_all((&new_uuid[..]).as_bytes())
        .expect("Unable to write uuid file");

    new_uuid
}

fn env_true_flag(
    cfg: &Holder<Holder<Enver, &ArgMatches, String, &str>, Localizer<Value>, String, String>,
    env_var: &str,
) -> bool {
    let m = cfg.grab().env(env_var).done();
    m.is_some() && m.unwrap() == "true"
}
